src/forbidden_tree.jl

```diff
@@ -0,0 +1,188 @@
+"""
+Forbids the subtree to be generated.
+A subtree is defined as a tree of `AbstractConstraintMatchNode`s. 
+Such a node can either be a `ConstraintMatchNode`, which contains a rule index corresponding to the 
+rule index in the grammar of the rulenode we are trying to match.
+It can also contain a `ConstraintMatchVar`, which contains a single identifier symbol.
+"""
+struct ForbiddenTree <: PropagatorConstraint
+	tree::AbstractConstraintMatchNode
+end
+
+
+# Matching RuleNode with ConstraintMatchNode
+"""
+Tries to match RuleNode `rn` with ConstraintMatchNode `cmn` and fill in 
+the domain of the hole at `hole_location`. 
+Returns if match is successful:
+
+- A dictionary with variable assignments
+- The id for the node which fills the hole
+A variable is represented by the index of its rulenode.
+Returns nothing if the match is unsuccessful.
+"""
+function _match_expr_with_hole(
+    rn::RuleNode, 
+    cmn::ConstraintMatchNode, 
+    hole_location::Vector{Int}
+)::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    hole_location == [] && throw(ArgumentError("The hole location doesn't point to a hole!"))
+
+    if rn.ind ≠ cmn.rule_ind || length(rn.children) ≠ length(cmn.children)
+        return nothing
+    else
+        vars = Dict()
+        forbidden_domain = nothing
+        for (i, rnᵢ, cmnᵢ) ∈ zip(1:length(rn.children), rn.children, cmn.children)
+            if i == hole_location[1]
+                # Match with finding domain of hole
+                forbidden_domain, varsᵢ = _match_expr_with_hole(rnᵢ, cmnᵢ, hole_location[begin+1:end])
+            else
+                # Regular match
+                varsᵢ = _match_expr(rnᵢ, cmnᵢ)
+            end
+
+            if varsᵢ ≡ nothing
+                # unsuccessful match
+                return nothing
+            end
+            # Check if another argument already assigned the same variables
+            for (k, v) ∈ varsᵢ
+                if k ∈ keys(vars) && v ≠ vars[k]
+                    return nothing
+                end
+                vars[k] = v
+            end
+        end
+        return forbidden_domain, vars
+    end
+end
+
+# Matching RuleNode with ConstraintMatchVar
+function _match_expr_with_hole(
```

Comment:
just for my own understnanding: MatchVar is not matched to RuleNode because the function is intended to work with holes?

Reply:
There are two versions of the match: `_match_expr` and `_match_expr_with_hole`.

- `_match_expr` just tries to match the RuleNode to the ConstraintMatchNode tree.
- `_match_expr_with_hole` explores the path to the hole, so the RuleNode that it gets is always a RuleNode that should have the hole we are expanding as one of its children. 

Since this is the `_match_expr_with_hole` function, we know that there will be a hole (the one we are expanding) in the rulenode. 
My initial reasoning was that assigning a tree with a hole to a variable will not work since when we check for equality of assigned values of identical variables when matching, we might get equality between two rulenode trees, but if they contain holes, they might get expanded to different things. Essentially, we cannot know yet whether they are equal. 

However, I think we can relax this a bit and return the empty domain in this case and match the expression to the variable. We then only have to check for holes when we combine variable assignments.

This other check for holes is something that I forgot to include, and actually causes a bug with larger examples, so thanks for asking this question!